astropy/io/fits/column.py

```diff
@@ -1521,6 +1523,11 @@ def _update_column_attribute_changed(self, column, attr, old_value,
             if col is column:
                 break
 
+        if attr == 'name':
+            del self.names
+        elif attr == 'format':
+            del self.formats
```

Comment:
Thanks Erik ! The column addition/removal is now fixed.

When you say `FITS_rec` class to be less reliant on the `ColDefs`, I wonder for example why ``FITS_rec.names` tries first to use `_coldefs.names` and then `dtype.names`, and not the opposite ? 

Based on our discussion here, I tried to make a few changes in `fitsrec.py`. This is on this [gist](https://gist.github.com/saimn/e2e6901be18cd24587202d776ac036d6), should I include this commit ?
- replace `self._coldefs.names` with `self.names`
- move indexing to `self._coldefs` when possible
- and change the way to get names using `dtype.names` first